### DISCUSSION TITLE
Disable lock file update for NuGet

### DISCUSSION THREAD

**Discussion by Otiel** [renovatebot/renovate#33000] [Suggest an Idea]
### Tell us more.

I'd like to create a PR in order to be able to disable the lock file update for NuGet. The documentation states that [`updatelockfiles`](https://docs.renovatebot.com/configuration-options/#updatelockfiles) config option is currently only supported for `npm`.

I suppose the only thing to do would be to add something along those lines at the start of the [`updateArtifacts`](https://github.com/renovatebot/renovate/blob/39.59.0/lib/modules/manager/nuget/artifacts.ts#L81) function:

```
if (!updateLockFiles) { // But how can I retrieve the value of updateLockFiles config option here?
  logger.debug('Skipping lock file update because disabled in config');
  return null;
}
```

But I'm having a hard time finding how the `updatelockfiles` config option is mapped to the objects available in the code.

1. Is this the right way to implement this PR?
2. How can I get the value of `updatelockfiles` config option from the the `updateArtifacts` function?

Thanks for your help.

**Comment by rarkins** [comment]
Why do you want to disable lock file updating? This is very old option for npm which will possibly be removed in future. Updating package files without lock files usually results in an invalid repository state.

**Reply by Otiel** [reply]
Lock file update for NuGet is done in Renovate by running `dotnet restore` individually on every `csproj` file. In my repository containing more than 200 `csprojs`, it takes around 30 minutes to run all the `dotnet restore` commands.

My plan was to disable the lock file update in Renovate but add a `postUpgradeTasks` such as `dotnet restore MySolution.sln` in order to update the lock files in a few seconds instead of 30 minutes, because I know that all my lock files are linked to `csprojs` that are included in my `sln`.

**Reply by rarkins** [reply]
Is this something which renovate could recognize and do automatically? Or too hard to recognize without help?

**Reply by Otiel** [reply]
I think this might be really tricky...

1. First because [`updateArtifacts`](https://github.com/renovatebot/renovate/blob/39.59.0/lib/modules/manager/nuget/artifacts.ts#L81) seems to be called by Renovate for each `csproj`, right?

    If so, we would have to detect if the `csproj` is part of a `sln` (that's the other tricky part - see below) before calling `dotnet restore TheCsprojSolution.sln`, then we would need to store the fact that `dotnet restore TheCsprojSolution.sln` has been called once in order not to call it again for the subsequent `csproj`s of that `sln`. Is this something conceivable in Renovate architecture?

2. Secondly, because I don't know of a sure way to detect if a `csproj` is part of a `sln`. To try to detect if a csproj is part of a `sln`, we could do:
    * Find every `sln` file in the repository. Is this feasible in Renovate?
    * For each `sln`, list its `csproj` files:
        * Either with [`dotnet sln MySolution.sln list` command](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-sln), then parse the output. Example of an output:
            ```
            λ dotnet sln MySolution.sln list
            Project(s)
            ----------
            Bcdl.Console\Bcdl.Console.csproj
            Bcdl.Downloader\Bcdl.Downloader.csproj
            Bcdl.IntegrationTests\Bcdl.IntegrationTests.csproj
            Bcdl.UI\Bcdl.UI.csproj
            Bcdl.UnitTests\Bcdl.UnitTests.csproj
            ```
            I'm a bit skeptical about parsing directly an output, but sadly `dotnet sln list` doesn't offer to save the result to a file. 😕
        * Or parse the `sln` file directly. But I don't like this, because here's an example of a `sln` content: 🤮
            ```
            Microsoft Visual Studio Solution File, Format Version 12.00
            # Visual Studio Version 16
            VisualStudioVersion = 16.0.31515.178
            MinimumVisualStudioVersion = 10.0.40219.1
            Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Bcdl.UnitTests", "Bcdl.UnitTests\Bcdl.UnitTests.csproj", "{DA30FD6E-5738-4892-A7DA-E4931A679BBB}"
            EndProject
            Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Bcdl.IntegrationTests", "Bcdl.IntegrationTests\Bcdl.IntegrationTests.csproj", "{DA2C3579-12D6-43FC-A4B5-5ABC4EAA0FC3}"
            EndProject
            Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Bcdl.UI", "Bcdl.UI\Bcdl.UI.csproj", "{A646B1F8-7567-44DC-91B8-6C4979444AB0}"
            EndProject
            Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Bcdl.Downloader", "Bcdl.Downloader\Bcdl.Downloader.csproj", "{510FB249-C006-4BDF-9F66-3F638D38AD5C}"
            EndProject
            Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Bcdl.Console", "Bcdl.Console\Bcdl.Console.csproj", "{6535F210-3AE6-46D1-A1BB-4B43A6FA883D}"
            EndProject
            Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "SolutionItems", "SolutionItems", "{8051AB5E-AC61-448A-A05A-30ADA36D6352}"
                ProjectSection(SolutionItems) = preProject
                    Directory.Build.props = Directory.Build.props
                EndProjectSection
            EndProject
            Global
                GlobalSection(SolutionConfigurationPlatforms) = preSolution
                    Debug|Any CPU = Debug|Any CPU
                    Release|Any CPU = Release|Any CPU
                EndGlobalSection
                GlobalSection(ProjectConfigurationPlatforms) = postSolution
                    {DA30FD6E-5738-4892-A7DA-E4931A679BBB}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
                    {DA30FD6E-5738-4892-A7DA-E4931A679BBB}.Debug|Any CPU.Build.0 = Debug|Any CPU
                    {DA30FD6E-5738-4892-A7DA-E4931A679BBB}.Release|Any CPU.ActiveCfg = Release|Any CPU
                    {DA30FD6E-5738-4892-A7DA-E4931A679BBB}.Release|Any CPU.Build.0 = Release|Any CPU
                    {DA2C3579-12D6-43FC-A4B5-5ABC4EAA0FC3}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
                    {DA2C3579-12D6-43FC-A4B5-5ABC4EAA0FC3}.Debug|Any CPU.Build.0 = Debug|Any CPU
                    {DA2C3579-12D6-43FC-A4B5-5ABC4EAA0FC3}.Release|Any CPU.ActiveCfg = Release|Any CPU
                    {DA2C3579-12D6-43FC-A4B5-5ABC4EAA0FC3}.Release|Any CPU.Build.0 = Release|Any CPU
                    {A646B1F8-7567-44DC-91B8-6C4979444AB0}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
                    {A646B1F8-7567-44DC-91B8-6C4979444AB0}.Debug|Any CPU.Build.0 = Debug|Any CPU
                    {A646B1F8-7567-44DC-91B8-6C4979444AB0}.Release|Any CPU.ActiveCfg = Release|Any CPU
                    {A646B1F8-7567-44DC-91B8-6C4979444AB0}.Release|Any CPU.Build.0 = Release|Any CPU
                    {510FB249-C006-4BDF-9F66-3F638D38AD5C}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
                    {510FB249-C006-4BDF-9F66-3F638D38AD5C}.Debug|Any CPU.Build.0 = Debug|Any CPU
                    {510FB249-C006-4BDF-9F66-3F638D38AD5C}.Release|Any CPU.ActiveCfg = Release|Any CPU
                    {510FB249-C006-4BDF-9F66-3F638D38AD5C}.Release|Any CPU.Build.0 = Release|Any CPU
                    {6535F210-3AE6-46D1-A1BB-4B43A6FA883D}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
                    {6535F210-3AE6-46D1-A1BB-4B43A6FA883D}.Debug|Any CPU.Build.0 = Debug|Any CPU
                    {6535F210-3AE6-46D1-A1BB-4B43A6FA883D}.Release|Any CPU.ActiveCfg = Release|Any CPU
                    {6535F210-3AE6-46D1-A1BB-4B43A6FA883D}.Release|Any CPU.Build.0 = Release|Any CPU
                EndGlobalSection
                GlobalSection(SolutionProperties) = preSolution
                    HideSolutionNode = FALSE
                EndGlobalSection
                GlobalSection(ExtensibilityGlobals) = postSolution
                    SolutionGuid = {C50CD3F6-F93D-48CC-8AE4-9E05B6094D42}
                EndGlobalSection
            EndGlobal
            ```
    * Find if the `csproj` is listed.

    But both ways to list the `csproj`s are not foolproof because we could imagine a repository with the following content (two `csproj`s with the same name, but included in a different `sln`):
    ```
    MySolution1\
        MySolution1.sln
        Bcdl.Console\Bcdl.Console.csproj
        Bcdl.Downloader\Bcdl.Downloader.csproj
        Bcdl.IntegrationTests\Bcdl.IntegrationTests.csproj
        Bcdl.UI\Bcdl.UI.csproj
        Bcdl.UnitTests\Bcdl.UnitTests.csproj
    MySolution2\
        MySolution2.sln
        Bcdl.Console\Bcdl.Console.csproj
    ```
    Then we would have no way of knowing if `Bcdl.Console.csproj` is part of `MySolution1.sln` or `MySolution2.sln`. 😟

All in all, this might prove hard to implement bug-free. What do you think?

**Reply by viceice** [reply]
the sln detection should be on extract phase and added to the package files manager object. then that information can be used for the update artifacts, something we do already for node based managers.

this probably needs some API changes, so extract phase can return the files on which the update artifacts should be called.

**Reply by Otiel** [reply]
I'm having a hard time finding the corresponding code to better understand what you mean. Can you point me to the file/function of the "extract phase" please?

Also, doing the `sln` detection during the extract phase would not solve the potential issue I described above with such a repository, would it?

```
MySolution1\
    MySolution1.sln
    Bcdl.Console\Bcdl.Console.csproj
    Bcdl.Downloader\Bcdl.Downloader.csproj
    Bcdl.IntegrationTests\Bcdl.IntegrationTests.csproj
    Bcdl.UI\Bcdl.UI.csproj
    Bcdl.UnitTests\Bcdl.UnitTests.csproj
MySolution2\
    MySolution2.sln
    Bcdl.Console\Bcdl.Console.csproj
```

**Reply by reduckted** [reply]
> Lock file update for NuGet is done in Renovate by running dotnet restore individually on every csproj file. In my repository containing more than 200 csprojs, it takes around 30 minutes to run all the dotnet restore commands.

This is the exact problem I'm having, except for me it's worse because I have 400+ projects, and it takes about an hour to restore the projects individually - it's so long that the GitHub app token I'm using actually expires before Renovate completes.

Has there been any progress on this?

**Reply by Otiel** [reply]
Nope, I need to implement #33118, but I lack time. 🙁

**Reply by reduckted** [reply]
#33118 has been completed, so this is now possible and works well.

```json
{
    "skipArtifactsUpdate": true,
    "postUpgradeTasks": {
        "commands": ["dotnet restore --force-evaluate --use-lock-file"],
        "executionMode": "branch",
        "fileFilters": ["**/packages.lock.json"]
    }
}
```

**Reply by Otiel** [reply]
@reduckted A huge thank you for implementing this! ❤️ 

I didn't have the time to implement it, but now thanks to you, I'll be able to configure and use Renovate in our air-gapped environment. I'm grateful for your work. 🙏

**Comment by rarkins** [comment]
FYI the original request can be satisfied once https://github.com/renovatebot/renovate/issues/33118 is implemented
